synapse/rest/client/v1/room.py

```diff
@@ -371,10 +372,23 @@ def __init__(self, hs):
             limit = None
 
         handler = self.hs.get_room_list_handler()
-        if server:
-            data = await handler.get_remote_public_room_list(
-                server, limit=limit, since_token=since_token
-            )
+        if server and server != self.hs.config.server_name:
+            try:
+                data = await handler.get_remote_public_room_list(
+                    server, limit=limit, since_token=since_token
+                )
+            except HttpResponseException as e:
+                if e.code == 401:
+                    raise SynapseError(
+                        400,
+                        "The remote server is only allowing local users to view "
+                        "its public rooms list",
+                        errcode=Codes.FORBIDDEN,
+                    )
+                else:
+                    raise SynapseError(
+                        400, "Unable to retrieve remote public rooms list"
+                    )
```

Comment:
can this go in `get_remote_public_room_list` rather than having two copies of it?